File: reference/dotnet/src/YawLabs.Awsp/Ssrf.cs
// AWSP -- A2A Webhook Security Profile reference implementation.
//
// Sender-side SSRF defense (SPEC.md section 10).
//
// A Receiver supplies its webhook URL during configuration. Without active defense, a hostile
// configuration call could point the URL at internal hosts (link-local metadata services, RFC
// 1918 networks, loopback) and trick the Sender into making requests on its behalf.
//
// This module implements the Sender-side checks the spec mandates:
//   1. Resolve the URL's hostname BEFORE connecting.
//   2. Reject the URL if any resolved address falls in a private, reserved, link-local,
//      multicast, or loopback range (full IPv4 + IPv6 list per SPEC.md section 10).
//   3. Return a URI with the host rewritten to the resolved IP literal so the caller can
//      open the connection by IP, defeating DNS-rebinding.
//   4. Refuse non-HTTPS schemes by default; opt-in for http:// via AllowHttp.
//
// Body-size caps, redirect-follow refusal, and total request time are caller responsibilities
// (configure them on HttpClient / HttpClientHandler).

using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;

namespace YawLabs.Awsp;

/// <summary>
/// Thrown by <see cref="Ssrf.AssertPublicUrlAsync"/> when a URL fails the SSRF gate.
/// </summary>
public class SsrfBlockedException : Exception
{
    /// <summary>Reason taxonomy for blocked URLs.</summary>
    public enum SsrfReason
    {
        /// <summary>The URL resolved to a private, reserved, link-local, multicast, or loopback address.</summary>
        PrivateIp,

        /// <summary>The input could not be parsed as an absolute URL with a host component.</summary>
        InvalidUrl,

        /// <summary>DNS resolution returned no usable addresses or threw.</summary>
        DnsFailure,

        /// <summary>The URL scheme is not allowed (https required by default; http opt-in via options).</summary>
        SchemeNotAllowed,
    }

    /// <summary>Reason this URL was blocked.</summary>
    public SsrfReason Reason { get; }

    /// <summary>The original URL string the caller passed in.</summary>
    public string Url { get; }

    /// <summary>The resolved IP that triggered the block, or <c>null</c> if the failure was pre-DNS.</summary>
    public string? ResolvedIp { get; }

    /// <summary>Construct a new SSRF block.</summary>
    public SsrfBlockedException(SsrfReason reason, string url, string? resolvedIp, string message)
        : base(message)
    {
        Reason = reason;
        Url = url;
        ResolvedIp = resolvedIp;
    }
}

/// <summary>
/// Options for <see cref="Ssrf.AssertPublicUrlAsync"/>.
/// </summary>
public class AssertPublicUrlOptions
{
    /// <summary>
    /// When <c>true</c>, http:// URLs are accepted in addition to https://. Default <c>false</c>.
    /// Per SPEC.md section 10, http MAY be permitted only on Sender-internal test fixtures or
    /// explicit operator opt-in.
    /// </summary>
    public bool AllowHttp { get; init; }

    /// <summary>
    /// Optional DNS resolver override. Defaults to <see cref="Dns.GetHostAddressesAsync(string, CancellationToken)"/>.
    /// Useful in tests to inject deterministic addresses without touching the network.
    /// </summary>
    public Func<string, CancellationToken, Task<IPAddress[]>>? Resolve { get; init; }
}

/// <summary>
/// Sender-side SSRF gate. Resolves a Receiver-supplied URL, rejects private/reserved IPs, and
/// returns a URI rewritten to the resolved public IP literal so the caller can connect by IP
/// and defeat DNS-rebinding.
/// </summary>
public static class Ssrf
{
    /// <summary>
    /// Resolve <paramref name="rawUrl"/>'s hostname, reject private/reserved IPs per SPEC.md
    /// section 10, and return a <see cref="Uri"/> with the host rewritten to the resolved
    /// public IP literal. Throws <see cref="SsrfBlockedException"/> on any failure.
    /// </summary>
    /// <param name="rawUrl">The URL to validate. Must be absolute with a host component.</param>
    /// <param name="opts">Options. <c>null</c> means defaults (https only, system DNS).</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>A URI identical to the input except the host is replaced with the resolved IP literal.</returns>
    /// <exception cref="SsrfBlockedException">The URL is invalid, uses a forbidden scheme, fails DNS, or resolves to a non-public address.</exception>
    public static async Task<Uri> AssertPublicUrlAsync(
        string rawUrl,
        AssertPublicUrlOptions? opts = null,
        CancellationToken ct = default)
    {
        opts ??= new AssertPublicUrlOptions();

        if (string.IsNullOrWhiteSpace(rawUrl))
        {
            throw new SsrfBlockedException(
                SsrfBlockedException.SsrfReason.InvalidUrl,
                rawUrl ?? string.Empty,
                null,
                "URL is null, empty, or whitespace.");
        }

        if (!Uri.TryCreate(rawUrl, UriKind.Absolute, out Uri? uri) || string.IsNullOrEmpty(uri.Host))
        {
            throw new SsrfBlockedException(
                SsrfBlockedException.SsrfReason.InvalidUrl,
                rawUrl,
                null,
                "URL is not a valid absolute URI with a host component.");
        }

        // Scheme gate. Default: https only. With AllowHttp: http or https.
        bool isHttps = string.Equals(uri.Scheme, Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase);
        bool isHttp = string.Equals(uri.Scheme, Uri.UriSchemeHttp, StringComparison.OrdinalIgnoreCase);
        if (!isHttps && !(isHttp && opts.AllowHttp))
        {
            throw new SsrfBlockedException(
                SsrfBlockedException.SsrfReason.SchemeNotAllowed,
                rawUrl,
                null,
                $"Scheme '{uri.Scheme}' is not allowed. Use https (or set AllowHttp = true for http).");
        }

        // Resolve. If the host is already an IP literal, IPAddress.TryParse handles it locally;
        // otherwise call out to DNS.
        IPAddress[] addresses;
        if (IPAddress.TryParse(uri.Host, out IPAddress? literal))
        {
            addresses = new[] { literal };
        }
        else
        {
            try
            {
                Func<string, CancellationToken, Task<IPAddress[]>> resolver =
                    opts.Resolve ?? ((host, token) => Dns.GetHostAddressesAsync(host, token));
                addresses = await resolver(uri.Host, ct).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                throw;
            }
            catch (Exception ex)
            {
                throw new SsrfBlockedException(
                    SsrfBlockedException.SsrfReason.DnsFailure,
                    rawUrl,
                    null,
                    $"DNS resolution failed for host '{uri.Host}': {ex.Message}");
            }

            if (addresses is null || addresses.Length == 0)
            {
                throw new SsrfBlockedException(
                    SsrfBlockedException.SsrfReason.DnsFailure,
                    rawUrl,
                    null,
                    $"DNS resolution returned no addresses for host '{uri.Host}'.");
            }
        }

        // Reject if ANY resolved address is non-public. The strictest interpretation: a single
        // private IP in the result set means we will not connect, even if a public one is also
        // present (a DNS-rebinding attacker can return a mix and steer the eventual connect
        // toward the private one).
        IPAddress? selected = null;
        foreach (IPAddress addr in addresses)
        {
            IPAddress canonical = NormalizeIpv4Mapped(addr);
            if (!IsPublicAddress(canonical))
            {
                throw new SsrfBlockedException(
                    SsrfBlockedException.SsrfReason.PrivateIp,
                    rawUrl,
                    canonical.ToString(),
                    $"Host '{uri.Host}' resolved to non-public address {canonical}.");
            }
            selected ??= canonical;
        }

        // Rewrite host to the resolved IP literal so the caller connects by IP.
        // IPv6 addresses must be wrapped in brackets in the URI.
        var builder = new UriBuilder(uri);
        if (selected!.AddressFamily == AddressFamily.InterNetworkV6)
        {
            builder.Host = "[" + selected + "]";
        }
        else
        {
            builder.Host = selected.ToString();
        }
        return builder.Uri;
    }

    // ----------------------------------------------------------------------
    // IP classification.
    // ----------------------------------------------------------------------

    /// <summary>
    /// True iff <paramref name="addr"/> is NOT in any of the private/reserved/link-local/
    /// multicast/loopback ranges enumerated in SPEC.md section 10.
    /// </summary>
    /// <remarks>
    /// Internal-visible for tests; the public surface intentionally exposes only the assert API.
    /// </remarks>
    internal static bool IsPublicAddress(IPAddress addr)
    {
        if (addr.AddressFamily == AddressFamily.InterNetwork)
        {
            return !IsPrivateIPv4(addr);
        }
        if (addr.AddressFamily == AddressFamily.InterNetworkV6)
        {
            return !IsPrivateIPv6(addr);
        }
        // Anything other than IPv4/IPv6 (e.g. AppleTalk, IPX) is not connectable here -- treat
        // as non-public for safety.
        return false;
    }

    private static IPAddress NormalizeIpv4Mapped(IPAddress addr)
    {
        // ::ffff:a.b.c.d -- the IPv4 rules apply per SPEC.md.
        if (addr.AddressFamily == AddressFamily.InterNetworkV6 && addr.IsIPv4MappedToIPv6)
        {
            return addr.MapToIPv4();
        }
        return addr;
    }

    private static bool IsPrivateIPv4(IPAddress addr)
    {
        byte[] b = addr.GetAddressBytes();

        // 0.0.0.0/8
        if (b[0] == 0) return true;
        // 10.0.0.0/8
        if (b[0] == 10) return true;
        // 100.64.0.0/10 (RFC 6598 carrier-grade NAT)
        if (b[0] == 100 && (b[1] & 0xC0) == 64) return true;
        // 127.0.0.0/8 loopback
        if (b[0] == 127) return true;
        // 169.254.0.0/16 link-local (incl. 169.254.169.254 cloud metadata)
        if (b[0] == 169 && b[1] == 254) return true;
        // 172.16.0.0/12 -- 172.16.0.0 through 172.31.255.255
        if (b[0] == 172 && (b[1] & 0xF0) == 16) return true;
        // 192.0.0.0/24 IETF protocol assignments
        if (b[0] == 192 && b[1] == 0 && b[2] == 0) return true;
        // 192.0.2.0/24 TEST-NET-1
        if (b[0] == 192 && b[1] == 0 && b[2] == 2) return true;
        // 192.168.0.0/16 private
        if (b[0] == 192 && b[1] == 168) return true;
        // 198.18.0.0/15 -- 198.18.0.0 through 198.19.255.255 (benchmarking)
        if (b[0] == 198 && (b[1] & 0xFE) == 18) return true;
        // 198.51.100.0/24 TEST-NET-2
        if (b[0] == 198 && b[1] == 51 && b[2] == 100) return true;
        // 203.0.113.0/24 TEST-NET-3
        if (b[0] == 203 && b[1] == 0 && b[2] == 113) return true;
        // 224.0.0.0/4 multicast (224 .. 239)
        if (b[0] >= 224 && b[0] <= 239) return true;
        // 240.0.0.0/4 reserved (240 .. 255), which subsumes 255.255.255.255/32
        if (b[0] >= 240) return true;

        return false;
    }

    private static bool IsPrivateIPv6(IPAddress addr)
    {
        byte[] b = addr.GetAddressBytes();
        if (b.Length != 16)
        {
            // Defensive -- should not happen for AddressFamily.InterNetworkV6.
            return true;
        }

        // ::/128 unspecified
        bool allZero = true;
        for (int i = 0; i < 16; i++)
        {
            if (b[i] != 0) { allZero = false; break; }
        }
        if (allZero) return true;

        // ::1/128 loopback
        bool isLoopback = true;
        for (int i = 0; i < 15; i++)
        {
            if (b[i] != 0) { isLoopback = false; break; }
        }
        if (isLoopback && b[15] == 1) return true;

        // ::ffff:0:0/96 IPv4-mapped. Already handled by NormalizeIpv4Mapped before reaching
        // here, but keep the guard so a raw IPv4-mapped IPAddress is still flagged.
        if (IsPrefixMatch(b, new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xFF, 0xFF }, 96))
        {
            return true;
        }

        // 64:ff9b::/96 well-known prefix for NAT64. Treat as non-public; if a Sender truly
        // wants NAT64 traversal it must explicitly target the embedded IPv4.
        if (IsPrefixMatch(b, new byte[] { 0x00, 0x64, 0xFF, 0x9B, 0, 0, 0, 0, 0, 0, 0, 0 }, 96))
        {
            return true;
        }

        // 100::/64 discard prefix
        if (IsPrefixMatch(b, new byte[] { 0x01, 0x00, 0, 0, 0, 0, 0, 0 }, 64))
        {
            return true;
        }

        // 2001::/23 IETF protocol assignments. Note: 2001:db8::/32 (documentation) is OUTSIDE
        // 2001::/23 (the /23 fixes bits 16-22 to zero; 2001:db8 has bit 20-21 set), so both
        // ranges are checked separately below.
        if (IsPrefixMatch(b, new byte[] { 0x20, 0x01, 0x00 }, 23))
        {
            return true;
        }

        // 2001:db8::/32 documentation prefix (disjoint from 2001::/23).
        if (IsPrefixMatch(b, new byte[] { 0x20, 0x01, 0x0D, 0xB8 }, 32))
        {
            return true;
        }

        // fc00::/7 unique local addresses (ULA)
        if ((b[0] & 0xFE) == 0xFC) return true;

        // fe80::/10 link-local
        if (b[0] == 0xFE && (b[1] & 0xC0) == 0x80) return true;

        // ff00::/8 multicast
        if (b[0] == 0xFF) return true;

        return false;
    }

    /// <summary>
    /// Compare the high <paramref name="prefixBits"/> of <paramref name="addr"/> against
    /// <paramref name="prefix"/>. Bits past the prefix in <paramref name="prefix"/> are ignored.
    /// </summary>
    private static bool IsPrefixMatch(byte[] addr, byte[] prefix, int prefixBits)
    {
        int fullBytes = prefixBits / 8;
        int remainderBits = prefixBits % 8;
        int compareBytes = Math.Min(fullBytes, prefix.Length);

        for (int i = 0; i < compareBytes; i++)
        {
            if (addr[i] != prefix[i])
            {
                return false;
            }
        }

        if (remainderBits == 0)
        {
            return true;
        }
        if (fullBytes >= prefix.Length || fullBytes >= addr.Length)
        {
            // Prefix is shorter than the bit count claims -- treat as no match for safety.
            return false;
        }
        int mask = (0xFF << (8 - remainderBits)) & 0xFF;
        return (addr[fullBytes] & mask) == (prefix[fullBytes] & mask);
    }
}

File: reference/python/yawlabs_awsp/_ssrf.py
"""Sender-side SSRF defense for Receiver-supplied webhook URLs.

SPEC.md section 10 mandates this. A Receiver supplies its webhook URL
during configuration; without active defense, an attacker controlling
that configuration can point the URL at internal hosts (cloud-metadata
service, RFC 1918 ranges, loopback) and trick the Sender into making
requests on its behalf.

`assert_public_url` resolves the hostname to one or more IP addresses,
rejects the URL if any resolved address falls in a private, reserved,
link-local, multicast, or loopback range, and returns a URL whose
hostname has been rewritten to the resolved public IP. Connecting to
the IP (not the hostname) defeats DNS-rebinding -- the IP we resolved
and gated is the IP we actually connect to.

Stdlib only: `ipaddress`, `socket`, `urllib.parse`. The default
resolver is `socket.getaddrinfo`; pass `resolve=` to inject a stub
for testing.

Caller pattern (requests / httpx):

    from yawlabs_awsp import assert_public_url, SsrfBlockedError

    try:
        safe_url = assert_public_url(receiver_url)
    except SsrfBlockedError as err:
        # log err.reason / err.url / err.resolved_ip; refuse to send
        raise

    # Connect by IP, send the original Host header so TLS / vhosting works.
    parsed = urllib.parse.urlparse(receiver_url)
    requests.post(safe_url, data=body, headers={"Host": parsed.netloc, ...})
"""

from __future__ import annotations

import ipaddress
import socket
from typing import Callable, List, Literal, Optional
from urllib.parse import urlparse, urlunparse

__all__ = ["SsrfBlockedError", "SsrfBlockReason", "assert_public_url"]


SsrfBlockReason = Literal[
    "private_ip",
    "invalid_url",
    "dns_failure",
    "scheme_not_allowed",
]


class SsrfBlockedError(Exception):
    """Raised when a URL fails SSRF gating.

    Attributes:
        reason: One of "private_ip", "invalid_url", "dns_failure",
            "scheme_not_allowed". Stable string suitable for logging /
            metrics.
        url: The URL that was rejected (the original input).
        resolved_ip: The offending IP, if applicable. Populated for
            "private_ip"; None for "invalid_url", "dns_failure",
            "scheme_not_allowed".
    """

    def __init__(
        self,
        reason: SsrfBlockReason,
        url: str,
        message: str,
        *,
        resolved_ip: Optional[str] = None,
    ) -> None:
        super().__init__(message)
        self.reason: SsrfBlockReason = reason
        self.url: str = url
        self.resolved_ip: Optional[str] = resolved_ip


# IPv4 ranges from SPEC.md section 10. Each entry is (network, label).
# Labels are descriptive, used only in error messages -- the SPEC list
# is authoritative.
_IPV4_BLOCKLIST: tuple[tuple[ipaddress.IPv4Network, str], ...] = (
    (ipaddress.IPv4Network("0.0.0.0/8"), "this-network"),
    (ipaddress.IPv4Network("10.0.0.0/8"), "rfc1918"),
    (ipaddress.IPv4Network("100.64.0.0/10"), "rfc6598-cgnat"),
    (ipaddress.IPv4Network("127.0.0.0/8"), "loopback"),
    (ipaddress.IPv4Network("169.254.0.0/16"), "link-local"),
    (ipaddress.IPv4Network("172.16.0.0/12"), "rfc1918"),
    (ipaddress.IPv4Network("192.0.0.0/24"), "ietf-protocol"),
    (ipaddress.IPv4Network("192.0.2.0/24"), "test-net-1"),
    (ipaddress.IPv4Network("192.168.0.0/16"), "rfc1918"),
    (ipaddress.IPv4Network("198.18.0.0/15"), "benchmark"),
    (ipaddress.IPv4Network("198.51.100.0/24"), "test-net-2"),
    (ipaddress.IPv4Network("203.0.113.0/24"), "test-net-3"),
    (ipaddress.IPv4Network("224.0.0.0/4"), "multicast"),
    (ipaddress.IPv4Network("240.0.0.0/4"), "reserved"),
    (ipaddress.IPv4Network("255.255.255.255/32"), "broadcast"),
)


# IPv6 ranges from SPEC.md section 10. ::ffff:0:0/96 (IPv4-mapped) is
# handled separately below: the embedded IPv4 is extracted and re-checked
# against _IPV4_BLOCKLIST so that, e.g., ::ffff:10.0.0.1 is rejected as
# rfc1918 rather than passing because IPv4-mapped is "just" reserved.
_IPV6_BLOCKLIST: tuple[tuple[ipaddress.IPv6Network, str], ...] = (
    (ipaddress.IPv6Network("::/128"), "unspecified"),
    (ipaddress.IPv6Network("::1/128"), "loopback"),
    (ipaddress.IPv6Network("64:ff9b::/96"), "nat64"),
    (ipaddress.IPv6Network("100::/64"), "discard-only"),
    (ipaddress.IPv6Network("2001::/23"), "ietf-protocol-assignments"),
    (ipaddress.IPv6Network("2001:db8::/32"), "documentation"),
    (ipaddress.IPv6Network("fc00::/7"), "unique-local"),
    (ipaddress.IPv6Network("fe80::/10"), "link-local"),
    (ipaddress.IPv6Network("ff00::/8"), "multicast"),
)


def _default_resolve(host: str) -> List[str]:
    """Default resolver: socket.getaddrinfo, returning all unique IPs."""
    seen: list[str] = []
    for info in socket.getaddrinfo(host, None):
        # info is (family, type, proto, canonname, sockaddr); sockaddr[0]
        # is the IP for both AF_INET and AF_INET6 (IPv6 sockaddr is a
        # 4-tuple but [0] is still the address).
        ip = info[4][0]
        # Strip IPv6 zone identifier if any (e.g. "fe80::1%eth0").
        if "%" in ip:
            ip = ip.split("%", 1)[0]
        if ip not in seen:
            seen.append(ip)
    return seen


def _check_ip(ip_str: str) -> Optional[str]:
    """Return a human label if `ip_str` is in any blocklist range, else None."""
    try:
        ip = ipaddress.ip_address(ip_str)
    except ValueError:
        # Not a parseable IP -- conservative default: block it. Can't
        # safely connect to something we can't even classify.
        return "unparseable-ip"

    if isinstance(ip, ipaddress.IPv4Address):
        for net, label in _IPV4_BLOCKLIST:
            if ip in net:
                return label
        return None

    # IPv6
    assert isinstance(ip, ipaddress.IPv6Address)

    # IPv4-mapped IPv6 (::ffff:0:0/96): check the embedded IPv4 against
    # the v4 blocklist. ipaddress.ip_address on the v4 form returns an
    # IPv4Address directly when we pass ip.ipv4_mapped.
    if ip.ipv4_mapped is not None:
        for net, label in _IPV4_BLOCKLIST:
            if ip.ipv4_mapped in net:
                return f"ipv4-mapped:{label}"
        return None

    for net, label in _IPV6_BLOCKLIST:
        if ip in net:
            return label
    return None


def assert_public_url(
    url: str,
    *,
    allow_http: bool = False,
    resolve: Optional[Callable[[str], List[str]]] = None,
) -> str:
    """Resolve a URL's hostname, reject private/reserved IPs, and return
    the URL with the hostname rewritten to the resolved public IP.

    The returned URL connects by IP (not hostname), defeating DNS-rebinding
    attacks. The original Host header SHOULD be preserved by the caller so
    TLS SNI and HTTP virtual hosting still work; see the module docstring
    for the requests/httpx pattern.

    Args:
        url: The URL to gate. MUST be https:// (or http:// if allow_http
            is True).
        allow_http: Permit http:// scheme. Default False -- spec section
            10 mandates HTTPS, with http:// reserved for sender-internal
            test fixtures or explicit operator opt-in.
        resolve: Optional hostname-to-IP-list resolver. Defaults to
            socket.getaddrinfo. Inject a stub for testing.

    Returns:
        A URL string with the hostname replaced by the first resolved
        public IP. IPv6 addresses are bracketed (e.g. https://[2606:...]/path).

    Raises:
        SsrfBlockedError: If the URL is malformed, the scheme is not
            permitted, DNS resolution fails, or any resolved IP is in
            the SPEC.md section 10 blocklist.
    """
    if not isinstance(url, str) or len(url) == 0:
        # Coerce non-string to string for the exception payload so
        # SsrfBlockedError.url stays well-typed.
        raise SsrfBlockedError(
            "invalid_url", url if isinstance(url, str) else "", "url must be a non-empty string"
        )

    try:
        parsed = urlparse(url)
    except ValueError as exc:
        raise SsrfBlockedError("invalid_url", url, f"url parse failed: {exc}") from exc

    scheme = (parsed.scheme or "").lower()
    if scheme == "https":
        pass
    elif scheme == "http":
        if not allow_http:
            raise SsrfBlockedError(
                "scheme_not_allowed",
                url,
                "http:// scheme not permitted; pass allow_http=True for test fixtures",
            )
    else:
        raise SsrfBlockedError(
            "scheme_not_allowed",
            url,
            f"scheme {scheme!r} not permitted; only https (or http with allow_http) allowed",
        )

    hostname = parsed.hostname
    if hostname is None or len(hostname) == 0:
        raise SsrfBlockedError("invalid_url", url, "url has no hostname")

    resolver = resolve if resolve is not None else _default_resolve

    try:
        ips = resolver(hostname)
    except Exception as exc:
        # socket.getaddrinfo raises socket.gaierror; stub resolvers may
        # raise anything. Treat all resolver failures uniformly.
        raise SsrfBlockedError(
            "dns_failure", url, f"dns resolution failed for {hostname!r}: {exc}"
        ) from exc

    if len(ips) == 0:
        raise SsrfBlockedError(
            "dns_failure", url, f"dns resolution returned no addresses for {hostname!r}"
        )

    for ip_str in ips:
        label = _check_ip(ip_str)
        if label is not None:
            raise SsrfBlockedError(
                "private_ip",
                url,
                f"resolved ip {ip_str} is in blocked range ({label})",
                resolved_ip=ip_str,
            )

    # Rewrite the URL's host to the first resolved public IP.
    target_ip = ips[0]
    try:
        target_obj = ipaddress.ip_address(target_ip)
    except ValueError:
        # Already filtered by _check_ip, so unreachable -- but be safe.
        raise SsrfBlockedError(
            "invalid_url", url, f"resolved ip {target_ip!r} is not parseable"
        )

    if isinstance(target_obj, ipaddress.IPv6Address):
        new_host = f"[{target_ip}]"
    else:
        new_host = target_ip

    new_netloc = new_host
    if parsed.port is not None:
        new_netloc = f"{new_host}:{parsed.port}"
    if parsed.username is not None:
        userinfo = parsed.username
        if parsed.password is not None:
            userinfo = f"{userinfo}:{parsed.password}"
        new_netloc = f"{userinfo}@{new_netloc}"

    return urlunparse(
        (
            parsed.scheme,
            new_netloc,
            parsed.path,
            parsed.params,
            parsed.query,
            parsed.fragment,
        )
    )
